### PR TITLE
Strip ANSI escape codes to compute cursor position

### LINKED ISSUE #135: Cursor Wonkiness When Using Console In Help Message
So, the project I've been working on is a little cli tool to play a game like [wordle](https://www.nytimes.com/games/wordle/index.html) right in the terminal.

I am using a [Text](https://github.com/mikaelmello/inquire#Text) input from the inquire crate to accept input from the user.

I'm also using the [console](https://crates.io/crates/console) crate to print out the remaining letters of the "keyboard" and highlight them to show if they have been guessed already and if they are in the word, correct position, etc.

I am also using some other crates for the generation of secret word and validation of guesses, but I don't think they are relevant to the console output wonkiness...

So, everything technically works right now, but the weird part is that when I start typing my whole prompt shifts up one line in my terminal. It's super weird and I'm not sure if the bug is caused by inquire, or console, or the two of them together, or the fact that I'm using all these things in the "help message" of my text input, or if it's just a unix thing...

It's hard to explain so I made this gif to try what I mean:

wonky-console-text-input

This project lives in the "wip_wordle_game" directory within [this repo](https://github.com/JimLynchCodes/rust-cli-examples/tree/wonky-console-output/wip_wordle_game). Feel free to take a look at the code and / or run it yourself with cargo run.

I'm looking for any advice on what I should try and / or insight into why this wonkiness is happening.

In the inquire docs it says, "You can choose your terminal backend between crossterm (default), termion or console".

I can try to reimplement the coloring using crossterm or termion, but I am very curious what's going on here with the console crate...

Thanks!

### PR DESCRIPTION
**Pull request by LeoRiether**
An attempt to fix #135

I added a module `crate::ansi` that parses ANSI escape codes according to [this description](https://handwiki.org/wiki/ANSI_escape_code#CSI_(Control_Sequence_Introducer)_sequences), which is then used to skip them when calculating the cursor position on `frame_finish()`;

Note that there are some rendering inconsistencies for ANSI codes between terminals, so it's impossible to make this work for every escape sequence in every terminal, but for non-pathological escape codes I think this implementation should be fine.
  
Pictured: inconsistencies between Kitty, Wezterm and Alacritty.
Screenshot_20230427_002742

Screenshot_20230427_005333

### CODE REVIEW

File: inquire/src/ansi.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+/// An iterator that strips ANSI escape codes from a string.
+///
+/// Generally constructed by calling [`strip_iter`].
+pub struct AnsiStripIter<'a> {
+    pub chars: Chars<'a>,
+}
+
+/// Constructs an iterator over the chars of the input string, stripping away ANSI escape codes.
```
Comment (mikaelmello) [inline inquire/src/ansi.rs:54]: Let's make a Trait with this method `into_ansi_strip_iter()` and implement it for &str

File: inquire/src/ui/backend.rs [hunk 2]
```diff
@@ -120,7 +120,7 @@ where
 
         let mut cur_pos = Position::default();
 
-        for (idx, c) in input.chars().enumerate() {
+        for (idx, c) in crate::ansi::strip_iter(input).enumerate() {
```
Comment (mikaelmello) [inline inquire/src/ui/backend.rs:123]: With the trait `input.into_ansi_strip_iter()` will look better

File: inquire/src/ansi.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+#[must_use]
+enum MatchResult<'a> {
+    Matched { remaining_chars: Chars<'a> },
+    NotMatched,
+}
+
+/// Match/consume the next ANSI escape code if it exists.
+/// The format is based on [this description](https://handwiki.org/wiki/ANSI_escape_code#CSI_(Control_Sequence_Introducer)_sequences)
+fn match_escape_code(chars: Chars<'_>) -> MatchResult<'_> {
```
Comment (mikaelmello) [inline inquire/src/ansi.rs:11]: Is this cross-platform? Should we use different rules depending on the platform? From a quick search looks like Windows ANSI and Standard ANSI differ

File: inquire/src/ansi.rs (hunk 1 above)
Comment (LeoRiether) [reply inquire/src/ansi.rs]: Good idea! I've implemented this in the last commit, but renamed it to `ansi_stripped_chars()` to more closely match `input.chars()` though, as this iterator is just a wrapper around `Chars`.

File: inquire/src/ansi.rs (hunk 3 above)
Comment (LeoRiether) [reply inquire/src/ansi.rs:11]: I think it is, but I'll need to research more before being sure.

**Review by mikaelmello — APPROVED** [review]
Only the changelog missing (and possibly the cross-platform double check)

**Comment by mikaelmello** [comment]
I've also confirmed this fixes #135 :)

**Comment by LeoRiether** [comment]
Cross-platform compatibility doesn't seem like a concern, from what I've seen. Terminal emulators like Alacritty, Wezterm and Kitty apparently use the same code to parse ANSI escape sequences, no matter the OS.

Cross-terminal compatibility, however, is impossible to achieve, so I'm not sure which standard to implement... Both Wezterm and Alacritty point to [this state machine](https://vt100.net/emu/dec_ansi_parser) for parsing escape and control sequences, but they're somehow inconsistent?? (this might be a bug in the "vte" crate)
more inconsistencies

I did overlook one thing in my implementation though: only CSI (the sequences that begin with `\x1b[` ~~or `\9b` in wezterm~~) are parsed currently. I should implement OSCs and DSCs too, in case someone wants to set their terminal's title in an inquire prompt I guess :)

**Comment by LeoRiether** [comment]
Should be mergeable now 👀

**Review by mikaelmello — APPROVED** [review]
Fantastic

**Pull request merged**